src/memray/_memray/inject.cpp

```diff
@@ -0,0 +1,245 @@
+#define Py_LIMITED_API 0x03070000
+#include "Python.h"
+
+#include <netdb.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <iostream>
+
+namespace memray {
+namespace {  // unnamed
+
+int
+connect_client(int port)
+{
+    struct addrinfo hints = {};
+    struct addrinfo* all_addresses = nullptr;
+    struct addrinfo* curr_address = nullptr;
+    int rv;
+    int sockfd;
+
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+
+    std::string port_str = std::to_string(port);
+    if ((rv = ::getaddrinfo(nullptr, port_str.c_str(), &hints, &all_addresses)) != 0) {
+        std::cerr << "getaddrinfo() failed while trying to attach Memray: " << ::gai_strerror(rv);
+        return -1;
+    }
+
+    for (curr_address = all_addresses; curr_address != nullptr; curr_address = curr_address->ai_next) {
+        sockfd = ::socket(curr_address->ai_family, curr_address->ai_socktype, curr_address->ai_protocol);
+        if (sockfd == -1) {
+            continue;
+        }
+
+        if (::connect(sockfd, curr_address->ai_addr, curr_address->ai_addrlen) == -1) {
+            ::close(sockfd);
+            continue;
+        }
+        break;
+    }
+    ::freeaddrinfo(all_addresses);
+    if (curr_address == nullptr) {
+        return -1;
+    }
+    return sockfd;
+}
+
+bool
+sendall(int fd, std::string_view data)
+{
+    size_t length = data.size();
+    while (length) {
+        ssize_t ret = ::send(fd, &data[0], length, 0);
```

Comment:
This can be `const`

Reply:
Sure, I guess, but I copy-pasted this out of `sink.cpp`. If it's important enough to you to ask me to change it here, we should probably go back and change it in `sink.cpp` as well.